#### David Felipe Mora - Ciencias de la computación
# Problemas


**PROBLEMA 1.** ¿Cuánta memoria requiere guardar 100.000 números de punto flotante de precisión doble?¿Cuánta memoria requiere guardar una matriz de números de punto flotante dimensión 100.000 x 100.000?.

**PROBLEMA 2.** Los números de punto flotante en python son implementados usando la clase float. Es decir, además de la descripción en bits del número tenemos asociados a la una instancia de variable tipo float los siguientes métodos y atributos para un dato float: as_integer_ratio, conjugate, fromhex, imag, is_integer, real .

Describa cada uno de estos métodos y atributos con ejemplos simples. 

La implementación del tipo de dato float en lenguaje C corresponde solo a la descripción en bits del número y no corresponde a una estructura u objeto. ¿Cuáles son las ventajas y desventajas de usar objetos?. 


**PROBLEMA 3.(Sullivan)** Con $x_0=1/10$, considere la sucesión,  
$$
x_{n+1}=\left\{\begin{array}{ll}
2x_n & \mbox{ si } x_n\in \left[0,\frac{1}{2}\right] \\
2x_n-1 &  \mbox{ si } x_n\in \left(\frac{1}{2},1\right].
\end{array}\right.
$$
para $n=1,2,\dots$. Calcule $x_1,x_2,x_3,\dots,x_{50}$ de forma análitica y despues usandoun código en python. Compare y explique.

Repetir pero ahora iniciando con $x_0=1/8$. Compare y explique. 

**PROBLEMA 4.** Explique la falla del Ariane. Muestre cálculos numéricos para soportar su explicación. 

**PROBLEMA 5.**  Explicar la falla del American Patriot. Muestre cálculos numéricos para soportar su explicación.

**PROBLEMA 6.** Ensaye calcular las funciones $$f(x)=\sqrt{x^2+1}-1 \quad g(x)=\frac{x^2}{\sqrt{x^2+1}+1}$$
Use valores pequeños como $x=8^{-1},8^{-2},8^{-3},\dots$. Explique los resultados. Note que, análiticamete,  $f(x)=g(x)$.

**PROBLEMA 7.(Quarteroni,Saleri,Gervasio)** El lector puede implementar el calculo de la siguiente recurrencia que aproxima $\pi$ (formula de Franc,ois Viète) y estudiar su estabilidad/inestabilidad numérica, 
$$z_1=2, \,\,z_{n+1}=2^{n-1/2}\sqrt{1-\sqrt{1-4^{1-n}z_n^2}},\,\,n=2,3,\dots. $$



**PROBLEMA 8.(Stewart)** Repita el experimento arriba con la recurrencia $x_1=\frac{1}{3}, x_2=\frac{1}{12}$,  $x_{k+1}=2.25x_k-0.5x_{k-1}$.

**PROBLEMA 9.** Presente un ejemplo, adicional  a los presentados en clase, de algoritmos inestables o problemas mal condicionados. 

**PROBLEMA 10.** Repita el experimento de arriba con $P(x)=(x-1)^7$. Agregue el siguiente método 
$$  P(x)= ((((((x-7)x+21)x-35)x+35)x-21)x+7)x-1 $$
Verifique esta formula. Implemente y comente los resultados. Esta forma de calcular polinomios es conocida como el esquema de Horner.

**PROBLEMA 11.(Stoer)** Considere los siguientes algoritmos. Algoritmo 1
$$s=p^2,\quad t=s+q,\quad u=\sqrt{t},\quad y=-p+u  $$
y el algoritmo 2, 
$$s=p^2,\quad t=s+q,\quad u=\sqrt{t}, \quad v=p+u,\quad y=q/v $$
Escriba $y$ como función de $p$ y $q$. Compare los resutados de estos dos algoritmos cuando $p=1000$,  $q=0.0018000000081$. Explique los resultados. ¿Cuál es mas confiable para estos valores de $p$ y $q$?

**PROBLEMA 12.** Considere el siguiente codigo
```
x=np.sqrt(1/9)
print("x={0:.60f}".format(x))
```
que dá como resultado 
x=0.333333333333333314829616256247390992939472198486328125000000.

Explique, tan detalladamente como le sea posible, lo que  está pasando. 
Compare con el siguiente código para el cual es más dificil saber la respuesta exacta. 
```
x=np.sqrt(0.1)
print("x={0:.60f}".format(x))
```
¿Cuántos decimales puede garantizar este cálculo?

**PROBLEMA 13.** Describa los siguientes tipos de datos :numpy.half,   numpy.float16, numpy.single, numpy.double, numpy.longdouble, numpy.csingle, numpy.cdouble, numpy.clongdouble.

# Soluciones

Las librerias que se van a usar son:

In [1]:
import math
import numpy as np

#### Problema 1.

Cualquier número de punto flotante en el formato de precision doble requiere 64 bits, o lo que es lo mismo, 8 Byte, luego guardar 100.000 de estos requiere 800 Kilobytes. Una matriz de 100.000 x 100.000 tiene $10^{10}$ números de punto flotante, luego ocuparía 80 Megabytes.

#### Problema 2.  

`x.as_integer_ratio()` retorna una tupla `(a, b)` donde $x = \frac{a}{b}$.

In [2]:
pi = math.pi
a, b = pi.as_integer_ratio()
rat = a / b
print(f"a = {a}")
print(f"b = {b}")
print(f"a/b = {rat}")

a = 884279719003555
b = 281474976710656
a/b = 3.141592653589793


`x.conjugate()` retorna el conjugado de `x`.

In [3]:
r = 5.0 # número real
print(f"Conjugado de 5.0: {r.conjugate()}")
c = 3 + 6j # número complejo 
print(f"Conjugado de 3 + 6j: {c.conjugate()}")

Conjugado de 5.0: 5.0
Conjugado de 3 + 6j: (3-6j)


`x.is_integer()` retorna `True` si `x` es entero.

In [4]:
print(f"¿Es pi entero?: {pi.is_integer()}")
print(f"¿Es 2.0 entero?: {2.0.is_integer()}")

¿Es pi entero?: False
¿Es 2.0 entero?: True


`x.real` es la parte real de `x`.

In [5]:
print(c)
print(c.real)

(3+6j)
3.0


`x.imag` es la parte compleja de `x`.

In [6]:
print(c)
print(c.imag)

(3+6j)
6.0


`float.fromhex(x)` donde `x` es un numero hexadecimal, retorna la representación decimal de `x` como `float`.

In [7]:
print(float.fromhex('0x3.a7p10'))

3740.0


El hecho que Python implementa en tipo de dato float mediante objetos permite proporcionar al usuario muchas funcionalidades como se ejemplificó anteriormente, sin embargo esta implementación mediante objetos podria tener penalizaciones en cuestiones de rendimiento en comparación a lenguajes como C ya que es una representación extra por la que se debe de pasar antes de hacer lo calculos.

#### Problema 3.

In [8]:
x_0 = 1/10
def f(x_n):
    if 0 <= x_n <= 1/2:
        return 2*x_n
    elif 1/2 < x_n <= 1:
        return 2*x_n - 1
    
x_n = x_0

for i in range(51):
    print(f"x_{i} = {x_n}")
    x_n = f(x_n)

x_0 = 0.1
x_1 = 0.2
x_2 = 0.4
x_3 = 0.8
x_4 = 0.6000000000000001
x_5 = 0.20000000000000018
x_6 = 0.40000000000000036
x_7 = 0.8000000000000007
x_8 = 0.6000000000000014
x_9 = 0.20000000000000284
x_10 = 0.4000000000000057
x_11 = 0.8000000000000114
x_12 = 0.6000000000000227
x_13 = 0.20000000000004547
x_14 = 0.40000000000009095
x_15 = 0.8000000000001819
x_16 = 0.6000000000003638
x_17 = 0.2000000000007276
x_18 = 0.4000000000014552
x_19 = 0.8000000000029104
x_20 = 0.6000000000058208
x_21 = 0.20000000001164153
x_22 = 0.40000000002328306
x_23 = 0.8000000000465661
x_24 = 0.6000000000931323
x_25 = 0.20000000018626451
x_26 = 0.40000000037252903
x_27 = 0.8000000007450581
x_28 = 0.6000000014901161
x_29 = 0.20000000298023224
x_30 = 0.4000000059604645
x_31 = 0.800000011920929
x_32 = 0.6000000238418579
x_33 = 0.20000004768371582
x_34 = 0.40000009536743164
x_35 = 0.8000001907348633
x_36 = 0.6000003814697266
x_37 = 0.20000076293945312
x_38 = 0.40000152587890625
x_39 = 0.8000030517578125
x_40 = 0.60000610

Analiticamente es claro que esta sucesion tiene un periodo de 4 despues del primer termino, por lo tanto es de la siguiente forma: 
$$(0.1, 0.2, 0.4, 0.8, 0.6, 0.2, 0.4, 0.8, 0.6, 0.2, 0.4, 0.8, 0.6, ...)$$

Comparando los resultados numericos contra los analiticos no damos cuenta que los numericos son cada vez mas imprecisos, esto posiblemente se debe a el error de redondedo que acarrea componer operaciones aritmeticas.

In [9]:
x_0 = 1/8
    
x_n = x_0
for i in range(50):
    print(f"x_{i} = {x_n}")
    x_n = f(x_n)

x_0 = 0.125
x_1 = 0.25
x_2 = 0.5
x_3 = 1.0
x_4 = 1.0
x_5 = 1.0
x_6 = 1.0
x_7 = 1.0
x_8 = 1.0
x_9 = 1.0
x_10 = 1.0
x_11 = 1.0
x_12 = 1.0
x_13 = 1.0
x_14 = 1.0
x_15 = 1.0
x_16 = 1.0
x_17 = 1.0
x_18 = 1.0
x_19 = 1.0
x_20 = 1.0
x_21 = 1.0
x_22 = 1.0
x_23 = 1.0
x_24 = 1.0
x_25 = 1.0
x_26 = 1.0
x_27 = 1.0
x_28 = 1.0
x_29 = 1.0
x_30 = 1.0
x_31 = 1.0
x_32 = 1.0
x_33 = 1.0
x_34 = 1.0
x_35 = 1.0
x_36 = 1.0
x_37 = 1.0
x_38 = 1.0
x_39 = 1.0
x_40 = 1.0
x_41 = 1.0
x_42 = 1.0
x_43 = 1.0
x_44 = 1.0
x_45 = 1.0
x_46 = 1.0
x_47 = 1.0
x_48 = 1.0
x_49 = 1.0


Analiticamente, es facil ver que los primeros 3 terminos son $0.125, 0.25, 0.5$ y lo que resta de la sucesion son $1.0$s.

En este caso los resultados numericos coinciden con los analiticos, esto es debido a que los terminos de la sucesion son todos numericos de maquina.

#### Problema 4.

El 4 de Junio de 1996, el lanzamiento del cohete **Ariane 5** explotó tan solo 40 segundos depues del despegue. Segun el reporte de la _Agencia Espacial Europea_, la principal causa del accidente fue una falla en el **sistema de control y orientación** de la aeronave, falla provocada por errores en el **diseño y especificaciones del software del sistema de referencia inercial**. En el [reporte de la mesa de investigación de la AEE](https://esamultimedia.esa.int/docs/esa-x-1819eng.pdf) se menciona especificamente que el error se dio al momento de convertir un número de punto flotante de **64 bits** a un entero de **16 bits**, el número que se estaba convirtiendo era demasiado grande y se produjo un **error por desbordamiento** en la forma de **_OPERAND ERROR_** que no había sido protegido en el codigo. 

Podemos mostrar la situación ocurrida mediante este simple ejemplo de codigo, donde aunque no se produce ningun error de conversión, las conversiones pierden sentido a partir de cierto punto:

In [10]:
for i in range(13):
    x = np.float64(10**i) # float64
    y = np.short(x) # int16
    print(f"float64:{x:15.1f} -> int16: {y}")

float64:            1.0 -> int16: 1
float64:           10.0 -> int16: 10
float64:          100.0 -> int16: 100
float64:         1000.0 -> int16: 1000
float64:        10000.0 -> int16: 10000
float64:       100000.0 -> int16: -31072
float64:      1000000.0 -> int16: 16960
float64:     10000000.0 -> int16: -27008
float64:    100000000.0 -> int16: -7936
float64:   1000000000.0 -> int16: -13824
float64:  10000000000.0 -> int16: 0
float64: 100000000000.0 -> int16: 0
float64:1000000000000.0 -> int16: 0


#### Problema 5.

El 25 de Febrero de 1991, una **sistema antimisiles _Patriot_** en Dhahran falló al momento de **seguir e interceptar** un misil _Scud_ debido a un problema del software del sistema de control que resultaba en **imprecisiones en los calculos de seguimiento** los cuales aumentaban a medida que lo hacia el tiempo de operacion del sistema. El sistema intentaba predecir la posición futura de un misil dada la velocidad y el tiempo de la ultima detección del mismo, el tiempo se guardaba internamente en números enteros que representan **decenas de segundos**. No obstante, a la hora de efecutar las predicciones el sistema requería el tiempo en segundos, por lo cual se debía escalar el tiempo por un factor de $\frac{1}{10}$. Los registros del sistema contaban con una capacidad de **24 bits**(No es claro en que formato), por lo que se **truncaban** los numeros pasados esta longitud de bits y se producia un **error de truncamiento**, teniendo en cuenta que el sistema llevaba unas **100 horas** en operación, se produjo un error de 0.34 segundos, lo suficiente para que el sistema no sea capaz de ubicar la posicion futura del misil con exactitud. Para mas información, consultar [IMTEC-92-26 Patriot Missile Defense: Software Problem Led to System Failure at Dhahran, Saudi Arabia](https://www.gao.gov/assets/imtec-92-26.pdf). A continuación se muestra el calculo de este error suponiendo que el formato usaba 20 bits para la mantisa para que los calculos sean conformes a los reales:

In [11]:
system_time = 100*60*60
time_in_sec = system_time * 10 # 100 horas en segundos 

error = 0.1/2**20

dec_error = "{:.9f}".format(error)
total_error = time_in_sec * error
print(f"Error absoluto por truncamiento: {error}")
print(f"Error en segundos: {total_error}")


Error absoluto por truncamiento: 9.5367431640625e-08
Error en segundos: 0.34332275390625


#### Problema 6.

In [12]:
def f(x):
    return math.sqrt(x**2 + 1) - 1

def g(x):
    return (x**2) / (math.sqrt(x**2 + 1) + 1)

for i in range(1, 15):
    x = 8**(-i)
    print(f"x = 1/8^{i} \t f(x) = {f(x):15.10e} \t g(x) = {g(x):15.10e}")

x = 1/8^1 	 f(x) = 7.7822185373e-03 	 g(x) = 7.7822185373e-03
x = 1/8^2 	 f(x) = 1.2206286283e-04 	 g(x) = 1.2206286283e-04
x = 1/8^3 	 f(x) = 1.9073468138e-06 	 g(x) = 1.9073468138e-06
x = 1/8^4 	 f(x) = 2.9802321944e-08 	 g(x) = 2.9802321944e-08
x = 1/8^5 	 f(x) = 4.6566128731e-10 	 g(x) = 4.6566128720e-10
x = 1/8^6 	 f(x) = 7.2759576142e-12 	 g(x) = 7.2759576142e-12
x = 1/8^7 	 f(x) = 1.1368683772e-13 	 g(x) = 1.1368683772e-13
x = 1/8^8 	 f(x) = 1.7763568394e-15 	 g(x) = 1.7763568394e-15
x = 1/8^9 	 f(x) = 0.0000000000e+00 	 g(x) = 2.7755575616e-17
x = 1/8^10 	 f(x) = 0.0000000000e+00 	 g(x) = 4.3368086899e-19
x = 1/8^11 	 f(x) = 0.0000000000e+00 	 g(x) = 6.7762635780e-21
x = 1/8^12 	 f(x) = 0.0000000000e+00 	 g(x) = 1.0587911841e-22
x = 1/8^13 	 f(x) = 0.0000000000e+00 	 g(x) = 1.6543612251e-24
x = 1/8^14 	 f(x) = 0.0000000000e+00 	 g(x) = 2.5849394142e-26


Se pude ver que los resultados de evaluar ambas funciones son aproximadamente los mismos hasta $8^{-8}$, despues de este punto los resultados de `f` son `0.0`, esto es debido a que hubo una perdida de digitos significativos mayor al numero de digitos de la mantisa provocado por restar dos digitos muy cercanos.

#### Problema 7.

In [13]:
z_1 = 2
def f(n, z_n):
    return 2**(n - 1/2) * math.sqrt(1 - math.sqrt(1 - (4**(1-n)) * (z_n**2)))
                                    
z_n = 2
print(f"z_1 = 2")
for n in range(2, 31):
    z_n = f(n, z_n)
    print(f"z_{n} = {z_n}")

z_1 = 2
z_2 = 2.8284271247461903
z_3 = 3.0614674589207187
z_4 = 3.121445152258053
z_5 = 3.136548490545941
z_6 = 3.140331156954739
z_7 = 3.141277250932757
z_8 = 3.1415138011441455
z_9 = 3.1415729403678827
z_10 = 3.141587725279961
z_11 = 3.141591421504635
z_12 = 3.141592345611077
z_13 = 3.141592576545005
z_14 = 3.1415926334632487
z_15 = 3.1415926548075896
z_16 = 3.1415926832667105
z_17 = 3.1415927591577
z_18 = 3.1415929109396727
z_19 = 3.141594125195191
z_20 = 3.1415965537048196
z_21 = 3.1415965537048196
z_22 = 3.1416742650217575
z_23 = 3.141829681889201
z_24 = 3.142451272494134
z_25 = 3.142451272494134
z_26 = 3.1622776601683795
z_27 = 3.1622776601683795
z_28 = 3.4641016151377544
z_29 = 4.000000000000001
z_30 = 8.000000000000002


Este algorimo es bastante inestable ya que aunque hasta las primeras 15 iteraciones el error disminuye, despues de este punto el error empieza a aumentar bastante rapido, esto posiblemente debido a el crecimiento del error de rendodeo producto de la composición de operaciones.

#### Problema 8.

In [14]:
x_1 = 1/3
x_2 = 1/12

def f(x_k, x_k_minus1):
    return 2.25*x_k - 0.5*x_k_minus1

x_3 = f(x_2, x_1)

x_k, x_k_minus1 = x_2, x_1

print(f"x_1 = {x_1}")
print(f"x_2 = {x_2}")
for i in range(10):
    x_k, x_k_minus_1 = f(x_k, x_k_minus1), x_k
    print(f"x_{i+3} = {x_k}")
    

x_1 = 0.3333333333333333
x_2 = 0.08333333333333333
x_3 = 0.020833333333333343
x_4 = -0.11979166666666663
x_5 = -0.4361979166666665
x_6 = -1.1481119791666663
x_7 = -2.7499186197916656
x_8 = -6.353983561197914
x_9 = -14.463129679361973
x_10 = -32.7087084452311
x_11 = -73.76126066843665
x_12 = -166.12950317064912


#### Problema 9.

Un ejemplo de un problema mal condicionado es el calculo de soluciones de sistemas de ecuaciones lineales que involucran la matriz: $$A^{T}A$$
Un ejemplo de este tipo de problemas son las ecuaciones normales que se usan para resolver problemas de optimización por minimos cuadrados: $$A^{T}Ax = A^{T}b$$ 
Debido a que el número de condición de $A^{T}A$ es aproximadamente el cuadrado del número de condición de $A$ se puede producir una aplimifacion importante el en error. Es por esto, que normalmente se opta por solucionar el sistema usando descomposición ortogonal como por ejemplo, descomposición QR.

#### Problema 10.

In [15]:
def f(x):
    return ((((((x-7)*x + 21)*x - 35)*x + 35)*x - 21)*x + 7)*x - 1


#### Problema 11. 

Las funciones para el algortimo 1 y 2 son respectivamente:
$$y_1(p, q) = \sqrt{p^2 + q} - p$$
$$y_2(p, q) = \frac{q}{\sqrt{p^2 + q} + p}$$


In [16]:
def y_1(p, q):
    return math.sqrt(p**2 + q) - q

def y_2(p, q):
    return q / (math.sqrt(p**2 + q) + p)

p = 1000
q = 0.001800000081
print(f'y_1 = {y_1(p,q)}')
print(f'y_2 = {y_2(p,q):.15f}')

y_1 = 999.9982008999191
y_2 = 0.000000900000040


En la primera formula que cuando $q \approx 0$:
$$y_1(p, q) \approx |p| - p$$
Por lo tanto tenemos una resta de numeros cercanos y por lo tanto se produce un error por perdida de digitos. Por lo tanto para los valores dados, es mejor utilizar el segundo algoritmo.